### DISCUSSION TITLE
Use hummingbird on iOS

### DISCUSSION THREAD

**Discussion by cflorion** [hummingbird-project/hummingbird#65] [Q&A] [answered]
Hi,

In our company we use Criollo (https://github.com/thecatalinstan/Criollo) in our iOS app, it acts as a server.
I just replaced it with Hummingbird and it seems to be working well, thanks a lot ! :)

1. Is this use (iOS) something you thought of ? Is it a good idea to use it in a production app or is it a hack in your opinion ?
2. Here is my implementation:
```
let address = HBBindAddress.hostname(localAddress, port: Int(port))
let configuration = HBApplication.Configuration(address: address)
let app = HBApplication(configuration: configuration)
app.start()
DispatchQueue.global(qos: .utility).async {
    app.wait()
}
```
If I don't put the `app.wait()`in another queue, the app freezes indefinitely (this is logical). Do you have any recommendation or using a global utility queue seems fine ?

Thanks!

**Comment by adam-fowler** [comment]
As it stands I wouldn't recommend you use Hummingbird on iOS. The current server backend uses POSIX sockets which you are discouraged from using on iOS. Your networking solutions for iOS should really be using Apple's `Network.framework`. Luckily the swift-nio team have built server components using this. See [swift-nio-transport-services](https://github.com/apple/swift-nio-transport-services). To get this working it would mean building a slightly different server backend. The code isn't that dissimilar so implementing it wouldn't be too hard. 

I was going to look at allowing for different server backends to be fitted in with Hummingbird. I was unhappy with the Lambda implementation I have at the moment and wanted to generalize the backend integration. I guess this gives me another reason.

**Reply by cflorion** [reply]
Thank you for such a clear answer! I have to study how SwiftNIO works so I could give you an hand on this.

About a Web Server on iOS, I think a good quality library like Hummingbird would be a great hit, because none of the existing ones are based on SwiftNIO, they use older dependencies like CocoaAsyncSocket.

Thanks!

**Reply by adam-fowler** [reply]
I have added a PR #71 which adds support for NIOTransportServices and therefore adds iOS as a supported platform. Would you be able to test it out.

**Reply by cflorion** [reply]
Wow! Of course !! I will do that ASAP
Thanks

**Reply by adam-fowler** [reply]
@cflorion have you had time to look at this yet?

**Reply by cflorion** [reply]
Hi @adam-fowler sorry I have not had the time yet, I planned to do it this weekend, I'll let you know, thanks!

**Reply by cflorion** [reply]
Hi @adam-fowler,

I have replaced our company main iOS app's HTTP layer with Hummingbird. I made a few tests and... It's working well !
Though I think one thing is missing to make it applicable to an iOS use. The `start()` method does not return if the starting was a success, if the iOS app needs to retry or not. For instance, when the app is goes in background I make sure to stop the server and I start it again when it goes back in foreground. So this is good. But, it happens that the port we ask for is already in use, and we loop over a list of ports until we find one free. So that is why we need to know how the start process went. Does that make sense ? Or maybe we can pass 0 as port and Hummingbird will pick the right port for us and give which one was picked ?

(I haven't implement the webSockets part yet)

The real test for us will be to put this version in production to our BETA users, but it won't be before a few weeks since we are testing other versions right now.

Thanks! The library was a pleasure to implement!

**Reply by adam-fowler** [reply]
The fact it isn't returning from `start` is a little concerning. 

Also is there a reason you are stopping the server when it goes into the background? Is it possible to just leave it running? Does that even work?

**Reply by cflorion** [reply]
Sorry my message was very unclear ! The method returns, what I meant is that I would need a return value, like ´Bool', or ´Error?´ for example to know if starting the server was a success or not. 
For example if you call start(), then later you start() again without stopping before. You get an error in the logs saying it failed starting because the port was already in use. So as an implementer, I don't know if calling start() the second time worked or not.

**Reply by cflorion** [reply]
About when going to background, iOS gives you a maximum of 30 seconds (iOS13) to perform your tasks and then it kills your processes, so the connection will be killed even if you don't close it yourself. So I prefer to close it cleanly my self. And when the app comes back in foreground I start the server again.

**Reply by adam-fowler** [reply]
Not sure why I wasn't throwing errors from `HBApplication.start`. Here is a PR that fixes that #73 
Yes going into background will close all connections.

**Reply by adam-fowler** [reply]
@cflorion All the changes are in release v0.11.0. Also I have been messing about with the iOS integration writing a small sample, that turns your phone into an image server (using your photo library as source). I'm pretty sure you don't need to stop and start the server when going into background and back. Just leave it running and everything seems to work fine.

**Comment by adam-fowler** [comment, marked as answer]
v0.11.0 now adds proper support for iOS

**Reply by cflorion** [reply]
That's great news ! I'm thinking about putting it to pour Beta users in production right now.The API of Hummingbird is really clean and I want to use it 🤩
I wanted your opinion on this, is Hummingbird's API quite stable or is it a bit too early to use it now ?

**Reply by adam-fowler** [reply]
The API is fairly stable. You will note it is still a v0.x.x so I cannot guarantee full stability, but I don't foresee any major changes. The core api won't change with the exception I will be adding async/await routes when that comes online.

Here is the iOS image server example I mentioned earlier. https://github.com/hummingbird-project/hummingbird-examples/pull/9

**Reply by cflorion** [reply]
Great, that enough for me! I'll put it to our beta testers and tell you how this goes ;)
Thanks a lot!

**Comment by cflorion** [comment]
Hi @adam-fowler 
I am struggling with setting up ```tlsOptions``` to have a HTTPS connection. Would you have a sample code or unit test that demonstrates how to use this please ?
Thanks a lot

**Reply by adam-fowler** [reply]
Ok what do you have so far? 
What is the address that you are connecting to the iPhone with? Is it just an IP addresss?
Also did you have tls working with Criollo?

**Reply by adam-fowler** [reply]
I must say I haven't given a huge amount of support here. The `NWProtocolTLS.Options` is quite an opaque interface. I can look to provide a easier to use interface, which takes a `SecIdentity` (you can create one of these using a `p12` file)

**Reply by cflorion** [reply]
I tried to do something with the information I found here https://developer.apple.com/forums/thread/113312. But I discarded what I did because I did not understand what I was doing. 

Yes, our current version that uses Criollo had HTTPS. With Criollo we just have to pass a certificate to the server and it becomes secure. https://github.com/thecatalinstan/Criollo#https

I'll read about SecIdentity, I don't know about it, this is interesting.
Thanks

**Reply by adam-fowler** [reply]
I have code in other repo's that generate a `NWProtocolTLS.Options`. I'll see if I can add another interface to `TSTLSOptions` which uses this.

**Reply by adam-fowler** [reply]
Ok I have a branch on HummingbirdCore `ts-tls-options` which has added an addition function for creating a `TSTLSOptions`. You can now create one with either
```swift
let tlsOptions = TSTLSOptions.options(serverIdentity: .p12(filename: p12Path, password: "p12Password"))
```
or
```swift
let tlsOptions = TSTLSOptions.options(serverIdentity: .secIdentity(MySecIdentity))
```
Can you test and verify this works for you.

**Reply by adam-fowler** [reply]
Just realised you'll need latest from the `main` branch of Hummingbird as well

**Reply by cflorion** [reply]
That is really clean thanks !
I just implemented it and it works very well

I had to fork the different dependencies (compression, webSocket) to make them point to the `main` branch of Hummingbird. Also I had to add `platforms: [.iOS(.v12), .tvOS(.v12)],` to their package.swift in order for them to compile.

Theses forks are a bit messy, I did that just to test it works. When you release a new version, I will do a cleaner implementation without forks and send a version with Hummingbird to our Beta users.

**Comment by adam-fowler** [comment]
I’ll get a release out and add the platforms to the Package.swift files

**Reply by adam-fowler** [reply]
Hummingbird v0.11.1 and HummingbirdCore v0.9.2 are out with the required changes.
I've also updated the other repo's to include the iOS and tvOS platform requirements

**Reply by cflorion** [reply]
Wow thanks !
